optuna/visualization/_terminator_improvement.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import annotations
+
+from typing import NamedTuple
+
+import tqdm
+
+import optuna
+from optuna._experimental import experimental_func
+from optuna.logging import get_logger
+from optuna.study.study import Study
+from optuna.terminator.erroreval import BaseErrorEvaluator
+from optuna.terminator.erroreval import CrossValidationErrorEvaluator
+from optuna.terminator.improvement.evaluator import BaseImprovementEvaluator
+from optuna.terminator.improvement.evaluator import DEFAULT_MIN_N_TRIALS
+from optuna.terminator.improvement.evaluator import RegretBoundEvaluator
+from optuna.visualization._plotly_imports import _imports
+
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go
+
+_logger = get_logger(__name__)
+
+
+PADDING_RATIO_Y = 0.05
+OPACITY = 0.25
+
+
+class _ImprovementInfo(NamedTuple):
+    trial_numbers: list[int]
+    improvements: list[float]
+    errors: list[float] | None
+
+
+@experimental_func("3.2.0")
+def plot_terminator_improvement(
+    study: Study,
+    plot_error: bool = False,
+    improvement_evaluator: BaseImprovementEvaluator | None = None,
+    error_evaluator: BaseErrorEvaluator | None = None,
+    min_n_trials: int = DEFAULT_MIN_N_TRIALS,
+) -> "go.Figure":
+    _imports.check()
+
+    info = _get_improvement_info(study, plot_error, improvement_evaluator, error_evaluator)
+    return _get_improvement_plot(info, min_n_trials)
+
+
+def _get_improvement_info(
+    study: Study,
+    get_error: bool = False,
+    improvement_evaluator: BaseImprovementEvaluator | None = None,
+    error_evaluator: BaseErrorEvaluator | None = None,
+) -> _ImprovementInfo:
+    if improvement_evaluator is None:
+        improvement_evaluator = RegretBoundEvaluator()
+    if error_evaluator is None:
+        error_evaluator = CrossValidationErrorEvaluator()
+
+    trial_numbers = []
+    improvements = []
+    errors = []
+
+    for i, trial in enumerate(tqdm.tqdm(study.trials)):
+        trial_numbers.append(trial.number)
+        trials = study.trials[: i + 1]
+        trials = [t for t in trials if t.state == optuna.trial.TrialState.COMPLETE]
```

Comment:
Does trial state change during this for loop? if not, the following change seems to make it faster and to provide the same results.
```suggestion
    trials = []
    for i, trial in enumerate(tqdm.tqdm(study.trials)):
        if trial.state == optuna.trial.TrialState.COMPLETE:
            trial_numbers.append(trial.number)
            trials.append(trial)
```

Reply:
I realized that the length of `trial_numbers` and `improvements` can be different in the original code, when study contains trials whose state is not complete. Is that on purpose?

Reply:
I think in the original code,  the length of `trial_numbers` and `improvements` is always same. Even if study contains trials whose state is not complete, the length of `improvements` is increased as the same timing as `trial_numbers`. The computation of `improvements` by `improvement_evaluator` is conducted only by completed trials, but it is on purpose.

Reply:
The first point you said is right. Thank you for your good catch! I updated my code according to your pointing.